trecs/models/mf.py

```diff
@@ -0,0 +1,294 @@
+            If True, enables verbose mode. Disabled by default.
+
+        num_items_per_iter: int (optional, default: 10)
+            Number of items presented to the user per iteration.
+
+        model_params: dict (optional, default: None)
+            Arguments that can be passed to `lenskit.algorithms.als.ImplicitMF()` that
+            dictate model training. For example: `{'iterations':40, 'reg':0.01}`.
+
+        seed: int, None (optional, default: None)
+            Seed for random generator.
+
+    Attributes
+    -----------
+        Inherited by BaseRecommender: :class:`~models.recommender.BaseRecommender`
+
+    Examples
+    ---------
+        ImplicitMF can be instantiated with no arguments -- in which case,
+        it will be initialized with the default parameters and the item/user
+        representations will be assigned randomly.
+
+        >>> mf = ImplicitMF()
+        >>> mf.users_hat.shape
+        (100, 10)   # <-- 100 users (default), 10 latent features (default)
+        >>> mf.items.shape
+        (10, 1250) # <-- 10 latent features (default), 1250 items (default)
+
+        This class can be customized either by defining the number of users/items/latent
+        features in the system.
+
+        >>> mf = ImplicitMF(num_users=1200, num_items=5000)
+        >>> mf.users_hat.shape
+        (1200, 10) # <-- 1200 users, 10 attributes
+
+        >>> mf = ImplicitMF(num_users=1200, num_items=5000, num_latent_features=2000)
+        >>> mf.users_hat.shape
+        (1200, 2000) # <-- 1200 users, 2000 attributes
+
+        Note that all arguments passed in at initialization must be consistent -
+        otherwise, an error is thrown. For example, one cannot pass in
+        `num_users=200` but have `user_representation.shape` be `(300, 100)`.
+        Likewise, one cannot pass in `num_items=1000` but have
+        `item_representation.shape` be `(100, 500)`.
+
+        To use the model, it's advised to use the `startup_and_train` method, which
+        will collect user interaction data from random recommendations. At the end
+        of the startup period, an implicit matrix factorization model is fit to the
+        interaction data, generating new user/item latent representations. These
+        representations are fixed and used for future calls to `run`.
+
+        >>> mf = ImplicitMF(num_latent_features=10) # 10 latent features
+        >>> mf.startup_and_train(50) # 50 timesteps of training
+        >>> mf.run(100)
+
+        If you'd like to retrain the matrix factorization model at any point, you
+        may use the `fit_mf` method. It will fit a new set of latent user/item
+        representations based on the interaction data from the *most recent* call to
+        `run()`.
+
+        >>> mf.run(50)
+        >>> mf.run(50) # on new call to `run`, interaction data from previous `run` is discarded
+        >>> mf.fit_mf() # MF model fit to data from most recent call to `run`
```

Comment:
My main hangup is the "departure" from the usual calling scheme. I am fine with `run()` not retraining at each step, as that is necessary iirc, but I'd rather avoid making it necessary to call `fit_mf()` explicitly. In other words, I would like fitting to happen seamlessly within `run()` to keep this model coherent with the execution scheme we use throughout the repo.

In the other models, we do:
```python
model.run(50, train_between_steps=False) # this trains after 50 steps
```
or:
```python
model.run(50, train_between_steps=True) # the default; this trains after each step for a total of 50 times
```
I understand this scheme is not sufficient for `ImplicitMF`. Can we extend it so that:
```python
modelMF.run(50, train_between_steps=False) # the default; this trains after 50 steps by calling fit_mf() internally?
modelMF.run(50, train_between_steps=True) # not allowed? See other comment I left about this
modelMF.run(50, no_training=True) # ignores the value of train_between_steps and does no training
```
... or something. I am not committed to adding a `no_training` parameter, nor am I committed on which should be the default for MF. What I want to do is change the scheme so that calling `fit_mf()` not be necessary. This will keep this model compatible with the rest of the models implemented so far.